cranelift/wasm/src/code_translator/bounds_checks.rs

```diff
@@ -1,132 +1,46 @@
-
-    pos.func
-        .dfg
-        .replace(inst)
-        .load(result_ty, flags, addr, Offset32::new(0));
-}
-
-/// Expand a `heap_store` instruction according to the definition of the heap.
-pub fn expand_heap_store(
-    inst: ir::Inst,
-    func: &mut ir::Function,
-    cfg: &mut ControlFlowGraph,
-    isa: &dyn TargetIsa,
-    heap_imm: ir::HeapImm,
-    index: ir::Value,
-    value: ir::Value,
-) {
-    let HeapImmData {
-        flags,
-        heap,
-        offset,
-    } = func.dfg.heap_imms[heap_imm];
-
-    let store_ty = func.dfg.value_type(value);
-    let access_size = u8::try_from(store_ty.bytes()).unwrap();
-
-    let mut pos = FuncCursor::new(func).at_inst(inst);
-    pos.use_srcloc(inst);
-
-    let addr =
-        bounds_check_and_compute_addr(&mut pos, cfg, isa, heap, index, offset.into(), access_size);
-
-    pos.func
-        .dfg
-        .replace(inst)
-        .store(flags, value, addr, Offset32::new(0));
-}
-
-/// Expand a `heap_addr` instruction according to the definition of the heap.
-pub fn expand_heap_addr(
-    inst: ir::Inst,
-    func: &mut ir::Function,
-    cfg: &mut ControlFlowGraph,
-    isa: &dyn TargetIsa,
-    heap: ir::Heap,
-    index: ir::Value,
-    offset: Uimm32,
-    access_size: Uimm8,
-) {
-    let mut pos = FuncCursor::new(func).at_inst(inst);
-    pos.use_srcloc(inst);
-
-    let addr =
-        bounds_check_and_compute_addr(&mut pos, cfg, isa, heap, index, offset.into(), access_size);
+//! Given
+//!
+//! * a dynamic Wasm memory index operand,
+//! * a static offset immediate, and
+//! * a static access size,
+//!
+//! bounds check the memory access and translate it into a native memory access.
 
-    // Replace the `heap_addr` and its result value with the legalized native
-    // address.
-    let addr_inst = pos.func.dfg.value_def(addr).unwrap_inst();
-    pos.func.dfg.replace_with_aliases(inst, addr_inst);
-    pos.func.layout.remove_inst(inst);
-}
+use crate::{HeapData, HeapStyle, TargetEnvironment};
+use cranelift_codegen::{
+    cursor::{Cursor, FuncCursor},
+    ir::{self, condcodes::IntCC, InstBuilder, RelSourceLoc},
+};
+use cranelift_frontend::FunctionBuilder;
 
 /// Helper used to emit bounds checks (as necessary) and compute the native
 /// address of a heap access.
 ///
 /// Returns the `ir::Value` holding the native address of the heap access.
-fn bounds_check_and_compute_addr(
-    pos: &mut FuncCursor,
-    cfg: &mut ControlFlowGraph,
-    isa: &dyn TargetIsa,
-    heap: ir::Heap,
+pub fn bounds_check_and_compute_addr<TE>(
+    builder: &mut FunctionBuilder,
+    env: &TE,
+    heap: &HeapData,
     // Dynamic operand indexing into the heap.
     index: ir::Value,
     // Static immediate added to the index.
     offset: u32,
     // Static size of the heap access.
     access_size: u8,
-) -> ir::Value {
-    let pointer_type = isa.pointer_type();
-    let spectre = isa.flags().enable_heap_access_spectre_mitigation();
```

Comment:
minor nit: I think the `spectre` local makes the cases a bit more readable below, vs. expanding out `env.enable_heap_access_spectre_mitigation()` everywhere, but I don't feel too strongly about it if you want to inline it...